File: agents/evals/lib/guardrails.mjs
// Reference implementation of instructions/guardrails.md, driven by guardrails.config.json.
// Used by the offline eval to exercise the real jailbreak patterns + PII rules. In production the
// Gradient console's built-in guardrails are configured to match the same JSON.
import { read } from "./util.mjs";

const cfg = JSON.parse(read("instructions/guardrails.config.json"));

const JAILBREAK_RES = cfg.jailbreak.patterns.map((p) => new RegExp(p, "i"));
const PII_RULES = cfg.anonymization.pii_types.map((t) => ({ ...t, re: new RegExp(t.regex, "g") }));

export function detectJailbreak(text) {
  const t = text || "";
  for (const re of JAILBREAK_RES) if (re.test(t)) return true;
  return false;
}

// Replace every PII match with its token. This is what runs BEFORE logging.
export function anonymize(text) {
  let out = text || "";
  const hits = [];
  for (const rule of PII_RULES) {
    out = out.replace(rule.re, () => { hits.push(rule.type); return rule.token; });
  }
  return { text: out, redacted_types: hits };
}

export const refusalMarkdown = cfg.jailbreak.refusal_markdown;

// Build the refusal envelope (valid §A) for a blocked request.
export function refusalEnvelope() {
  return {
    agent: "spot_scout",
    reply_markdown: refusalMarkdown,
    citations: [],
    map_actions: [],
    checklist: null
  };
}

File: backend/app/guardrails.py
"""Guardrails ported from agents/instructions/guardrails.config.json.

Same semantics as agents/evals/lib/guardrails.mjs: jailbreak patterns refuse
with a valid §A envelope; PII is anonymized BEFORE anything is logged. The
config JSON stays the single source of truth — this module only interprets it.
"""

from __future__ import annotations

import json
import re
from functools import lru_cache
from typing import Any

from .settings import get_settings


@lru_cache(maxsize=1)
def _config() -> dict[str, Any]:
    path = get_settings().instructions_dir / "guardrails.config.json"
    return json.loads(path.read_text(encoding="utf-8"))


@lru_cache(maxsize=1)
def _jailbreak_patterns() -> list[re.Pattern[str]]:
    return [re.compile(p, re.IGNORECASE) for p in _config()["jailbreak"]["patterns"]]


@lru_cache(maxsize=1)
def _pii_rules() -> list[tuple[str, str, re.Pattern[str]]]:
    return [
        (rule["type"], rule["token"], re.compile(rule["regex"]))
        for rule in _config()["anonymization"]["pii_types"]
    ]


def detect_jailbreak(text: str | None) -> bool:
    value = text or ""
    return any(pattern.search(value) for pattern in _jailbreak_patterns())


def anonymize(text: str | None) -> tuple[str, list[str]]:
    """Replace every PII match with its token; returns (text, redacted_types)."""
    out = text or ""
    hits: list[str] = []
    for pii_type, token, pattern in _pii_rules():
        def _sub(_match: re.Match[str], t: str = pii_type, tok: str = token) -> str:
            hits.append(t)
            return tok

        out = pattern.sub(_sub, out)
    return out, hits


def refusal_markdown() -> str:
    return _config()["jailbreak"]["refusal_markdown"]


def refusal_envelope(agent: str = "spot_scout") -> dict[str, Any]:
    return {
        "agent": agent,
        "reply_markdown": refusal_markdown(),
        "citations": [],
        "map_actions": [],
        "checklist": None,
    }
